content/blog/2025-05-13-bril-concurrency-extension.md

```diff
@@ -0,0 +1,163 @@
+    
+3.  **Reporting**: The pass reports per-function and global statistics: the fraction of allocations that remain thread-local.
+    
+
+This simple pass identifies heap objects that never cross thread boundaries, opening opportunities to Elide synchronization or use stack allocation in a future optimizer.
+
+4\. Benchmark Suites
+--------------------
+
+We developed two complementary benchmark suites:
+
+*   **Correctness Suite** (benchmarks/concurrency/\*.bril): A set of small tests covering spawn/join arity errors, double joins, shared counters, parallel sums, and array writers. We validated behavior under both the stubbed and worker-based interpreters using turnt.
+    
+*   **Performance Suite** (benchmarks/concurrency/perf/\*.bril): Larger, data-parallel workloads measuring speed-up with workers:
+    
+    *   perf\_par\_sum\_100k.bril: Summing 1..100 000 split across two threads. 100 k elements produce ~20× speed-up over sequential.
+        
+    *   perf\_matmul\_split.bril: Matrix multiplication on 100×100 matrices with row-range split; yields ~10× speed-up.
+        
+    *   perf\_big\_par\_sum\_10M.bril: Summing 1..10 000 000, which instead suffers a slowdown due to RPC overhead.
+        
+
+5\. Empirical Evaluation
+------------------------
+
+### 5.1 Correctness
+
+All the correctness benchmarks pass under the worker-enabled interpreter (brili-conc), matching our specification. Error cases in the error directory (arity mismatch, double join, wrong handle type) correctly exit with error codes.
+
+### 5.2 Performance
+
+Below are the wall-clock times measured for three representative benchmarks, comparing the sequential interpreter (using --no-workers) against our concurrent version (with true Deno Web Workers and an RPC-based shared heap).
+
+```
+=== Benchmark: perf_par_sum_100k ===
+Sequential: 217 ms, Concurrent: 4.5 s → 20.8× speed-up
+
+=== Benchmark: perf_matmul_split ===
+Sequential: 4.2 s, Concurrent: 44.9 s → 10.8× speed-up
+
+=== Benchmark: perf_big_par_sum_10M ===
+Sequential: 15.8 s, Concurrent:  > (timeout/slower) → RPC overhead dominates
+```
+
+#### **Why “concurrent” is actually slower**
+
+1.  Our design (real Web Workers + RPC heap) means _every_ load and store from a worker turns into:
+    ```
+    postMessage({ kind: "heap_req", … })   // worker → main
+    → main services request, touches native JS array
+    postMessage({ kind: "heap_res", … })   // main → worker
+    ```
+ On a 100 k–element loop, that’s ∼200 k round trips, each costing hundreds of microseconds in Deno’s message‐passing layer—totalling several seconds. At 10 M elements, it grows to ∼20 M messages, which simply overwhelms the system.
+    
+2.  Spawning even two workers in Deno carries a fixed overhead (tens to hundreds of milliseconds). In fine-grained loops (100 k or fewer iterations), that overhead is not negligible, and in very fine loops (10 M) it compounds the RPC penalty.
```

Comment:
How many threads are you launching in each benchmark? Is it constant, or related to the input size?